### PR TITLE
[Source-MsSQL] Set default cursor for cdc mode

### PR DESCRIPTION
**Pull request by nguyenaiden**
## What
To prepare for Destination v2, Cdc-enabled DB sources need to have a pre-defined cursor for normalization to operate properly.

## How
image
During the discover phase, `source_defined_cursor` is set to `true`, and a new composite cursor column called `_ab_cdc_cursor` is selected as the predefined cursor. This column is constructed from:

1. Converting the emittedAt timestamp , generated at the [read step](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-relational-db/src/main/java/io/airbyte/integrations/source/relationaldb/AbstractDbSource.java#L143), to epoch seconds.
2. Convert that to nanoseconds.
3. Initialize a threadsafe recordCounter = 1
4. Add the recordCounter to the converted timestamp
5. Increment the counter

The numeric cursor value will be computed for every record and allows for fast comparison in tie-breaking scenarios for records with the same Primary Key.
```
Updated Streamstate
{
  "cdc": false,
  "streams": [
    {
      "stream_name": "users",
      "cursor_field": [
        "_ab_cdc_cursor"
      ],
      "stream_namespace": "dbo"
    }
  ],
```

## Recommended reading order
1. `MsSQLSource.java`
6. `MsSqlCdcConnectorMetadataInjector.java`
7.  All the test files

## 🚨 User Impact 🚨
New CDC syncs that have refreshed their source schema will see this cursor field be chosen.
This update requires that users do a full reset **OR** drop their SCD tables to ensure that syncs continue operating normally
Customers who reset their data will continue to see old behavior until their source goes through the `discover()` phase again.

### CODE REVIEW

**Comment by octavia-squidington-iii** [comment]
##  source-mssql test report (commit [`c6f9dee920`](https://github.com/airbytehq/airbyte/pull/29493/commits/c6f9dee920c9203d084fd7ee635e4ad3052a79bf)) - ❌

⏲️  Total pipeline duration: 20mn03s 

| Step                                                                    | Result   |
|:------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-mssql/metadata.yaml     | ✅        |
| Connector version semver check                                          | ✅        |
| Connector version increment check                                       | ❌        |
| QA checks                                                               | ✅        |
| Build connector tar                                                     | ✅        |
| Build source-mssql docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-mssql:integrationTest | ❌        |
| Acceptance tests                                                        | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/sql-server-cdc-cursor/1692209935/c6f9dee920c9203d084fd7ee635e4ad3052a79bf/source-mssql/1.1.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:c6f9dee920c9203d084fd7ee635e4ad3052a79bf)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql test
```

**Comment by octavia-squidington-iii** [comment]
##  source-mssql test report (commit [`c6f9dee920`](https://github.com/airbytehq/airbyte/pull/29493/commits/c6f9dee920c9203d084fd7ee635e4ad3052a79bf)) - ❌

⏲️  Total pipeline duration: 18mn48s 

| Step                                                                    | Result   |
|:------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-mssql/metadata.yaml     | ✅        |
| Connector version semver check                                          | ✅        |
| Connector version increment check                                       | ❌        |
| QA checks                                                               | ✅        |
| Build connector tar                                                     | ✅        |
| Build source-mssql docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-mssql:integrationTest | ❌        |
| Acceptance tests                                                        | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/sql-server-cdc-cursor/1692211239/c6f9dee920c9203d084fd7ee635e4ad3052a79bf/source-mssql/1.1.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:c6f9dee920c9203d084fd7ee635e4ad3052a79bf)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql test
```

**Comment by octavia-squidington-iii** [comment]
##  source-mssql test report (commit [`b056c83f60`](https://github.com/airbytehq/airbyte/pull/29493/commits/b056c83f60c0c298722d6c1645cd3b7a3576350f)) - ❌

⏲️  Total pipeline duration: 20mn24s 

| Step                                                                    | Result   |
|:------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-mssql/metadata.yaml     | ✅        |
| Connector version semver check                                          | ✅        |
| Connector version increment check                                       | ❌        |
| QA checks                                                               | ✅        |
| Build connector tar                                                     | ✅        |
| Build source-mssql docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-mssql:integrationTest | ❌        |
| Acceptance tests                                                        | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/sql-server-cdc-cursor/1692212958/b056c83f60c0c298722d6c1645cd3b7a3576350f/source-mssql/1.1.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:b056c83f60c0c298722d6c1645cd3b7a3576350f)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql test
```

**Comment by octavia-squidington-iii** [comment]
##  source-mssql test report (commit [`67e3ec3cc1`](https://github.com/airbytehq/airbyte/pull/29493/commits/67e3ec3cc18c6cf24952a2426919a97849fe4234)) - ❌

⏲️  Total pipeline duration: 20mn51s 

| Step                                                                    | Result   |
|:------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-mssql/metadata.yaml     | ✅        |
| Connector version semver check                                          | ✅        |
| Connector version increment check                                       | ❌        |
| QA checks                                                               | ✅        |
| Build connector tar                                                     | ✅        |
| Build source-mssql docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-mssql:integrationTest | ❌        |
| Acceptance tests                                                        | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/sql-server-cdc-cursor/1692218564/67e3ec3cc18c6cf24952a2426919a97849fe4234/source-mssql/1.1.1/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:67e3ec3cc18c6cf24952a2426919a97849fe4234)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql test
```

**Comment by octavia-squidington-iii** [comment]
##  source-mssql test report (commit [`bc733328b1`](https://github.com/airbytehq/airbyte/pull/29493/commits/bc733328b1e1d76b0a4a49fb857ebb84f8258351)) - ❌

⏲️  Total pipeline duration: 20mn27s 

| Step                                                                    | Result   |
|:------------------------------------------------------------------------|:---------|
| Validate airbyte-integrations/connectors/source-mssql/metadata.yaml     | ❌        |
| Connector version semver check                                          | ✅        |
| Connector version increment check                                       | ✅        |
| QA checks                                                               | ✅        |
| Build connector tar                                                     | ✅        |
| Build source-mssql docker image for platform linux/x86_64               | ✅        |
| ./gradlew :airbyte-integrations:connectors:source-mssql:integrationTest | ❌        |
| Acceptance tests                                                        | ✅        |

🔗 [View the logs here](https://storage.cloud.google.com/airbyte-ci-reports-multi/airbyte-ci/connectors/test/pull_request/sql-server-cdc-cursor/1692219964/bc733328b1e1d76b0a4a49fb857ebb84f8258351/source-mssql/2.0.0/output.html)

☁️ [View runs for commit in Dagger Cloud](https://alpha.dagger.cloud/changeByPipelines?filter=dagger.io/git.ref:bc733328b1e1d76b0a4a49fb857ebb84f8258351)

*Please note that tests are only run on PR ready for review. Please set your PR to draft mode to not flood the CI engine and upstream service on following commits.*
**You can run the same pipeline locally on this branch with the [airbyte-ci](https://github.com/airbytehq/airbyte/blob/master/airbyte-ci/connector_ops/connector_ops/pipelines/README.md) tool with the following command**
```bash
airbyte-ci connectors --name=source-mssql test
```

**Review by prateekmukhedkar — COMMENTED** [review]
I believe we need to do the changes for `source-mssql-strict-encrypt` also.

File: airbyte-integrations/connectors/source-mssql/Dockerfile
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/Dockerfile:27]: Do you also want to do this for mssql-strict-encrypt?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:459]: One quick thought I had was : 
if this is now stateful, do we want to declare this in the constructor to prevent other parts of the code from declaring multiple versions of the CdcConnectorMetadataInjector?

**Review by akashkulk — APPROVED** [review]
Approving to unblock. Some quick nits

**Review by akashkulk — CHANGES_REQUESTED** [review]
Actually probably want to look into test failures before merging

File: airbyte-integrations/connectors/source-mssql/Dockerfile
Comment (nguyenaiden) [reply airbyte-integrations/connectors/source-mssql/Dockerfile:27]: Good call. Will add that

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (nguyenaiden) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:459]: @akashkulk  I actually already turned it into a singleton. Refer to `MssqlCdcConnectorMetadataInjector.java`

**Review by akashkulk — APPROVED** [review]

**Comment by nguyenaiden** [comment]
Successful test runs:
`source-mssql`: https://github.com/airbytehq/airbyte/actions/runs/5954575782/job/16151423518
`source-mssql-strict-encrypt`: https://github.com/airbytehq/airbyte/actions/runs/5954220427/job/16150814124

**Comment by octavia-approvington** [comment]
After a careful ML study, <br/> we think this looks okay.  <br/> imagine code being okay

**Review by octavia-approvington — APPROVED** [review]

**Pull request merged**
